ext/twig/twig.c

```diff
@@ -0,0 +1,758 @@
+{
+	int success;
+	zval retval;
+
+	if (arg2) {
+		zend_call_method(object, Z_OBJCE_P(object), NULL, method, strlen(method), &retval, 2, arg1, arg2);
+	} else {
+		zend_call_method(object, Z_OBJCE_P(object), NULL, method, strlen(method), &retval, 1, arg1, NULL);
+	}
+
+	success = Z_TYPE(retval) == IS_TRUE;
+
+	if (result) {
+		ZVAL_COPY(result, &retval);
+	}
+
+	zval_ptr_dtor(&retval);
+
+	return success;
+}
+
+static int twig_has_extension(zval *obj)
+{
+	zval str;
+	int ret;
+
+	ZVAL_STR_COPY(&str, twig_extension_sandbox_str);
+
+	ret = twig_call_method(obj, "hasextension", NULL, &str, NULL);
+
+	zval_ptr_dtor(&str);
+
+	return ret;
+}
+
+static int twig_add_array_key_to_string(zval *pDest, int num_args, va_list args, zend_hash_key *hash_key)
+{
+	smart_string *buf = va_arg(args, smart_string*);
+
+	if (buf->len != 0) {
+		smart_string_appends(buf, ", ");
+	}
+
+	if (!hash_key->key) {
+		smart_string_append_long(buf, hash_key->h);
+	} else {
+		zend_string *key, *tmp_str;
+		key     = php_addcslashes(hash_key->key, 0, "'\\", 2);
+		tmp_str = php_str_to_str(ZSTR_VAL(key), ZSTR_LEN(key), "\0", 1, "' . \"\\0\" . '", 12);
+
+		smart_string_appendl(buf, ZSTR_VAL(tmp_str), ZSTR_LEN(tmp_str));
+		zend_string_release(key);
+		zend_string_release(tmp_str);
+	}
+
+	return ZEND_HASH_APPLY_KEEP;
+}
+
+static char *twig_implode_array_keys(zval *array)
+{
+	smart_string collector = { 0, 0, 0 };
+
+	zend_hash_apply_with_arguments(HASH_OF(array), twig_add_array_key_to_string, 1, &collector);
+	smart_string_0(&collector);
+
+	return collector.c;
+}
+
+static void twig_runtime_error(zval *source, char *message, ...)
+{
+	zend_string *buffer;
+	va_list args;
+	zend_fcall_info_cache fcic = {0};
+	zend_fcall_info fci        = {0};
+	zval constructor_retval, ex, constructor_args[3];
+
+	va_start(args, message);
+	buffer = vstrpprintf(0, message, args);
+	ZEND_ASSERT(buffer);
+	va_end(args);
+
+	object_init_ex(&ex, TWIG_G_R(twig_error_runtime_ce));
+
+	ZVAL_STR(&constructor_args[0], buffer);
+	ZVAL_LONG(&constructor_args[1], -1);
+
+	fci.size           = sizeof(fci);
+	fci.params         = constructor_args;
+	fci.param_count    = 3;
+	fci.retval         = &constructor_retval;
+
+	fcic.called_scope     = TWIG_G_R(twig_error_runtime_ce);
+	fcic.calling_scope    = TWIG_G_R(twig_error_runtime_ce);
+	fcic.function_handler = TWIG_G_R(twig_error_runtime_ce)->constructor;
+	fcic.object           = Z_OBJ(ex);
+	fcic.initialized      = 1;
+
+	ZVAL_COPY_VALUE(&constructor_args[2], source);
+
+	zend_call_function(&fci, &fcic);
+
+	zval_ptr_dtor(&constructor_args[1]);
+	zval_ptr_dtor(&constructor_retval);
+
+	Z_ADDREF(ex);
+	zend_throw_exception_object(&ex);
+}
+
+/* {{{ proto mixed twig_get_attribute(Twig_Environment env, Twig_Source source, mixed object, mixed item, array arguments, string type, boolean isDefinedTest, boolean ignoreStrictCheck)
+   A C implementation of TwigTemplate::getAttribute() */
```

Comment:
should be updated, as the PHP code does not live in ``TwigTemplate::getAttribute()`` anymore